### PR TITLE
revert modifiers definition

### PR DESCRIPTION
**Pull request by pmazzini**
Revert the change to the modifiers so that we can have a modifier that encapsulates packets.

The PR in question is https://github.com/insomniacslk/dhcp/pull/253.

### CODE REVIEW

File: dhcpv6/modifiers.go [hunk 1]
```diff
@@ -10,26 +10,28 @@ import (
 
 // WithClientID adds a client ID option to a DHCPv6 packet
 func WithClientID(duid Duid) Modifier {
-	return func(d DHCPv6) {
+	return func(d DHCPv6) DHCPv6 {
```
Comment (hugelgupf) [inline dhcpv6/modifiers.go]: :/

This means we can't use specific types with modifiers, or you have to ignore the return value.

I feel like encapsulation should not be done by a modifier. Modifiers as they are here are used to add stuff to the inside of the packet, not to redefine the packet itself. I'd just call that an Encapsulator and make it a separate thing?

File: dhcpv6/modifiers_test.go [hunk 2]
```diff
@@ -46,32 +46,30 @@ func TestWithRequestedOptions(t *testing.T) {
...
-	WithIANA(OptIAAddress{
+	d := WithIANA(OptIAAddress{
 		IPv6Addr:          net.ParseIP("::1"),
 		PreferredLifetime: 3600,
 		ValidLifetime:     5200,
-	})(&d)
-	require.Equal(t, 1, len(d.Options))
-	require.Equal(t, OptionIANA, d.Options[0].Code())
+	})(&Message{})
+	require.Equal(t, 1, len(d.(*Message).Options))
```
Comment (hugelgupf) [inline dhcpv6/modifiers_test.go]: Now you have to do all these casts, and that's crummy. I'd prefer to separate Modifiers and Encaps rather than reintroduce the need for casting.

Comment (hugelgupf) [reply dhcpv6/modifiers_test.go]: In my mind, casting like this will only lead to production panics and reboots.

File: dhcpv6/modifiers.go (hunk 1 above)
Comment (insomniacslk) [reply dhcpv6/modifiers.go]: originally the modifiers were intended to be able to also modify the packet if necessary, but I agree that this is not the right way to use them

Comment (pmazzini) [reply dhcpv6/modifiers.go]: > I'd just call that an Encapsulator and make it a separate thing?

What about the client code? We send Relay messages using modifiers.

File: dhcpv6/modifiers_test.go (hunk 2 above)
Comment (pmazzini) [reply dhcpv6/modifiers_test.go]: > Now you have to do all these casts

I am reverting as few things as possible to get some internal code to work again.
These casts were previously there.

> In my mind, casting like this will only lead to production panics and reboots.

I agree but these are only tests.

File: dhcpv6/modifiers.go (hunk 1 above)
Comment (pmazzini) [reply dhcpv6/modifiers.go]: I am not against it but until we come up with a solution, I would revert it.
This is what we are currenty using to do unicast tests to specific servers.

File: dhcpv6/modifiers_test.go (hunk 2 above)
Comment (hugelgupf) [reply dhcpv6/modifiers_test.go]: > > Now you have to do all these casts
> 
> I am reverting as few things as possible to get some internal code to work again.
> These casts were previously there.
> 
> > In my mind, casting like this will only lead to production panics and reboots.
> 
> I agree but these are only tests.

There is no such thing as "only tests". These tests use exported features of the library. That means any user that wants to use the library has to cast like this to make it work. Yes, it was previously there, but that's the reason for the improvement.

File: dhcpv6/modifiers.go (hunk 1 above)
Comment (hugelgupf) [reply dhcpv6/modifiers.go]: > I am not against it but until we come up with a solution, I would revert it.
> This is what we are currenty using to do unicast tests to specific servers.

Can you wait to import the new code until you come up with a solution?

I'd love to come up with one for you, but without seeing the code, it's hard to say what would be a good way to deal with it.

Comment (pmazzini) [reply dhcpv6/modifiers.go]: > Can you wait to import the new code until you come up with a solution?

Sure!

> I'd love to come up with one for you, but without seeing the code, it's hard to say what would be a good way to deal with it.

We have a modifier similar to:

```
func encapsulateRelay(peerAddr net.IP) dhcpv6.Modifier {
	return func(packet dhcpv6.DHCPv6) dhcpv6.DHCPv6 {
		encapsulatedPacket, err := dhcpv6.EncapsulateRelay(packet, dhcpv6.MessageTypeRelayForward, net.IPv6zero, peerAddr)
		return encapsulatedPacket
	}
}
```

Which is passed to the client's `Exchange` method.

So the client creates a Solicit, applies the modifier and sends a RelayForw.
Then it receives a RelayRepl and decapsulates it:
https://github.com/insomniacslk/dhcp/blob/master/dhcpv6/client6/client.go#L61

With the advertise, the client creates a request, applies the modifier and sends again a RelayForw.

So basically with 1 modifier we were emulating a relay, which broke after changing how modifiers behave.

What we need is a client supporting exchange of relay messages, which we were doing with a modifier.

Comment (hugelgupf) [reply dhcpv6/modifiers.go]: > func encapsulateRelay(peerAddr net.IP) dhcpv6.Modifier {
> 	return func(packet dhcpv6.DHCPv6) dhcpv6.DHCPv6 {
> 		encapsulatedPacket, err := dhcpv6.EncapsulateRelay(packet, dhcpv6.MessageTypeRelayForward, net.IPv6zero, peerAddr)
> 		return encapsulatedPacket
> 	}
> }

I feel like this is kind of a dangerous modifier, because it's the only modifier I've seen where it's ordering with /other/ modifiers matters. E.g.

```go
client.Exchange(ifname, encapsulateRelay(nil), WithClientID(foobar))
```
is a different packet than
```go
client.Exchange(ifname, msg, WithClientID(foobar), encapsulateRelay(nil))
```

And I feel this is not intuitive. I think It wouldn't be that hard to do something like
```go
client.Exchange(ifname, Encap(msg, WithClientID(foo)))
```
instead, and it's *very clear* which message the modifier applies to in this model.

This would also mean that you can drop the modifiers on client.Exchange. It might be useful to have a helper like this instead:

```go
func Apply(msg DHCPv6, modifiers ...Modifier) DHCPv6

func Encap(msg DHCPv6, modifiers ...Modifier) *RelayMessage

func EncapWithFoo(...) *RelayMessage
```

and then you can
```go
client.Exchange(ifname, Apply(msg, WithClientID(foo)))
client.Exchange(ifname, Apply(Encap(msg, WithClientID(bar)), WithClientID(foo)))
client.Exchange(ifname, Encap(Apply(msg, WithClientID(bar)))
```

I'm also confused about clients that act like relays. This is just for testing, right? Because I imagine a relay is actually made using the `server6` package rather than `client6`.

Comment (pmazzini) [reply dhcpv6/modifiers.go]: > I feel like this is kind of a dangerous modifier, because it's the only modifier I've seen where it's ordering with /other/ modifiers matters. E.g.

Well, the order of the rest of the modifiers also matters, since we have a list of options instead of a map, it affects the ordering of the options.

> and then you can

```
client.Exchange(ifname, Apply(msg, WithClientID(foo)))
```

Ok, this means changing the client API so that `Exchange` can now take `DHCPv6` packets.

```
client.Exchange(ifname, Encap(Apply(msg, WithClientID(bar)))
```

How is the Request generated in this case by the client? I mean, how is it encapsulated again?

> This is just for testing, right?

No, for doing checks to servers (monitoring).

Comment (hugelgupf) [reply dhcpv6/modifiers.go]: > > I feel like this is kind of a dangerous modifier, because it's the only modifier I've seen where it's ordering with /other/ modifiers matters. E.g.
> 
> Well, the order of the rest of the modifiers also matters, since we have a list of options instead of a map, it affects the ordering of the options.

I feel the same about that. The order should not matter. The marshaling order of options *should* not affect DHCP clients & servers either, and it doesn't per spec. Modifiers should be applicable in any order.

For encapsulation, it's not the marshaling order that's affected, but *which* packet the modifier gets used on. Is it an option for the encapsulated message, or the relay message? Depends on whether it's before or after the relay modifier. That's too subtle for me from an API design perspective.

> 
> > and then you can
> 
> ```
> client.Exchange(ifname, Apply(msg, WithClientID(foo)))
> ```
> Ok, this means changing the client API so that `Exchange` can now take `DHCPv6` packets.
> 
> ```
> client.Exchange(ifname, Encap(Apply(msg, WithClientID(bar)))
> ```
> How is the Request generated in this case by the client? I mean, how is it encapsulated again?

Ah, sorry, you're right. I think for something like this, a specialized client like yours should build its Relay message the right way, and then just use SendAndRead to send the packet. I.e. I think to send a different kind of packet, you should not use Exchange.

> 
> > This is just for testing, right?
> 
> No, for doing checks to servers (monitoring).

Comment (pmazzini) [reply dhcpv6/modifiers.go]: How about adding the `PeerAddr`attribute to the client? Whenever that is set, the packet gets entcapsulated before sending it, that should solve our use case.

Comment (pmazzini) [reply dhcpv6/modifiers.go]: It should be a simple change.

**Comment by pmazzini** [comment]
Sorry for the noise. I am going to put up different PR to fix this.
